src/unix.rs

```diff
@@ -46,16 +49,47 @@ impl AsFd for UdpSocket {
     }
 }
 
+#[derive(Clone, Debug)]
+pub(crate) struct LastSendError(Arc<AtomicU64>);
+
+impl Default for LastSendError {
+    fn default() -> Self {
+        let now = Self::now();
+        Self(Arc::new(AtomicU64::new(
+            now.checked_sub(2 * IO_ERROR_LOG_INTERVAL).unwrap_or(now),
+        )))
+    }
+}
+
+impl LastSendError {
+    fn now() -> u64 {
+        SystemTime::now()
+            .duration_since(SystemTime::UNIX_EPOCH)
+            .unwrap()
+            .as_secs()
+    }
+
+    // Determine whether the last error was more than IO_ERROR_LOG_INTERVAL
+    // seconds ago. If so, update the last error time and return true.
+    pub(crate) fn should_log(&self) -> bool {
+        let now = Self::now();
+        self.0
+            .fetch_update(Ordering::SeqCst, Ordering::SeqCst, |cur| {
+                ((now - cur) > IO_ERROR_LOG_INTERVAL).then_some(now)
```

Comment:
any concerns here since the now is not guaranteed to be > cur ? Because it's systemtime and no monotonic, an NTP update could roll this back and get us a panic, no?

Reply:
Good point.  Given how infrequent that is these days, do you think a `saturating_sub` would fit the bill, and you just accept that you might miss one cycle of messages in that unlikely event?

Reply:
looks good, thanks!